blockchain/store_test.go

```diff
@@ -0,0 +1,269 @@
+
+	// End of setup, test data
+
+	tuples := []struct {
+		block            *types.Block
+		parts            *types.PartSet
+		seenCommit       *types.Commit
+		wantErr          bool
+		wantPanic        bool
+		eraseDBForCommit bool
+	}{
+		{
+			block: &types.Block{
+				Header: &types.Header{
+					Height:  1,
+					NumTxs:  100,
+					ChainID: "block_test",
+				},
+				LastCommit: &types.Commit{Precommits: []*types.Vote{{Height: 10}}},
+			},
+			parts:      validPartSet,
+			seenCommit: seenCommit1,
+		},
+
+		{
+			block:     nil,
+			wantPanic: true, // want a panic on saving nil block
+		},
+
+		{
+			block: &types.Block{
+				Header: &types.Header{
+					Height:  4,
+					NumTxs:  100,
+					ChainID: "block_test",
+				},
+				LastCommit: &types.Commit{Precommits: []*types.Vote{{Height: 10}}},
+			},
+			parts:     uncontiguousPartSet,
+			wantPanic: true, // and incomplete and uncontiguous parts
+		},
+
+		{
+			block: &types.Block{
+				Header: &types.Header{
+					Height:  1,
+					NumTxs:  100,
+					ChainID: "block_test",
+				},
+				LastCommit: &types.Commit{Precommits: []*types.Vote{{Height: 10}}},
+			},
+			parts:     incompletePartSet,
+			wantPanic: true, // incomplete parts
+		},
+
+		{
+			block: &types.Block{
+				Header: &types.Header{
+					Height:  1,
+					NumTxs:  100,
+					ChainID: "block_test",
+				},
+				LastCommit: &types.Commit{Precommits: []*types.Vote{{Height: 10}}},
+			},
+			parts:            validPartSet,
+			seenCommit:       seenCommit1,
+			eraseDBForCommit: true,
+			wantPanic:        true, // erase all content so that the commit will not be read
+		},
+	}
+
+	type quad struct {
+		block  *types.Block
+		commit *types.Commit
+		meta   *types.BlockMeta
+
+		seenCommit *types.Commit
+	}
+
+	for i, tuple := range tuples {
+		bs, db := freshBlockStore()
+		// SaveBlock
+		res, err, panicErr := doFn(func() (interface{}, error) {
+			bs.SaveBlock(tuple.block, tuple.parts, tuple.seenCommit)
+			if tuple.block == nil {
+				return nil, nil
+			}
+			bBlock := bs.LoadBlock(tuple.block.Height)
+			bBlockMeta := bs.LoadBlockMeta(tuple.block.Height)
+			bSeenCommit := bs.LoadSeenCommit(tuple.block.Height)
+			if tuple.eraseDBForCommit {
+				eraseDB(db)
+			}
+			bCommit := bs.LoadBlockCommit(tuple.block.Height - 1)
+			return &quad{block: bBlock, seenCommit: bSeenCommit, commit: bCommit, meta: bBlockMeta}, nil
+		})
+
+		if tuple.wantPanic {
+			if panicErr == nil {
+				t.Errorf("#%d: want a non-nil panic", i)
+			}
+			continue
+		}
+
+		if tuple.wantErr {
+			if err == nil {
+				t.Errorf("#%d: got nil error", i)
+			}
+			continue
+		}
+
+		assert.Nil(t, panicErr, "#%d: unexpected panic", i)
+		assert.Nil(t, err, "#%d: expecting a non-nil error", i)
+		qua, ok := res.(*quad)
+		if !ok || qua == nil {
+			t.Errorf("#%d: got nil quad back; gotType=%T", i, res)
+			continue
+		}
+	}
+}
+
+func doFn(fn func() (interface{}, error)) (res interface{}, err error, panicErr error) {
```

Comment:
is this worth it? why not just let it panic in the test?

Reply:
I believe we need this function. The point of it is to capture the panic and relay it back to whatever caller needs to check if we expected a panic or not -- for example if the blockMeta content is messed up vs when a block is successfully read.
If we just let the test panic, we wouldn't be asserting expected behavior and the tests wouldn't matter. I hope this makes sense.